## Parts 1, 2



In [1]:
import requests
import pandas as pd
from bs4 import BeautifulSoup

response = requests.get("https://worldpopulationreview.com/state-rankings/starbucks-stores-by-state")
soup = BeautifulSoup(response.content, "html.parser")

table = soup.find("table")
headers = [header.text.strip() for header in table.find_all("th")]
headers = headers[:3]
rows = []
state_names = []

for row in table.find_all("tr")[1:]:
    state_name = row.find("th").find("a").text.strip()
    state_names.append(state_name)
    cells = [cell.text.strip() for cell in row.find_all("td")]
    rows.append(cells)

df_s = pd.DataFrame(rows, columns=headers)
df_s.index = state_names

In [2]:
df_s = df_s.rename(columns={'State': 'starbucks_locations'})
df_s = df_s.drop(['Starbucks Stores 2023', 'Starbucks Stores 2021'], axis=1)

In [3]:
df_s['State'] = df_s.index
df_s = df_s.reset_index(drop=True)

In [4]:
df_s

starbucks_locations                 State
0                3,080            California
1                1,346                 Texas
2                  844               Florida
3                  741            Washington
4                  692              New York
5                  677              Illinois
6                  548               Arizona
7                  495              Colorado
8                  491                  Ohio
9                  489              Virginia
10                 359                Oregon
11                 357          Pennsylvania
12                 338        North Carolina
13                 326               Georgia
14                 283              Michigan
15                 273         Massachusetts
16                 261            New Jersey
17                 257              Maryland
18                 253                Nevada
19                 221               Indiana
20                 188              Missouri
21                 184             Minnesota
22                 180             Tennessee
23                 166              Kentucky
24                 145             Wisconsin
25                 131        South Carolina
26                 123           Connecticut
27                 101                  Utah
28                  99                Hawaii
29                  94                Kansas
30                  91  District of Columbia
31                  89                  Iowa
32                  85               Alabama
33                  84             Louisiana
34                  79              Oklahoma
35                  76            New Mexico
36                  67                 Idaho
37                  58              Nebraska
38                  55              Arkansas
39                  49                Alaska
40                  32           Mississippi
41                  32               Montana
42                  30                 Maine
43                  29         New Hampshire
44                  27          Rhode Island
45                  25              Delaware
46                  25          South Dakota
47                  25         West Virginia
48                  23               Wyoming
49                  13          North Dakota
50                   8               Vermont

In [5]:
response1 = requests.get("https://worldpopulationreview.com/state-rankings/dunkin-donuts-by-state")
soup1 = BeautifulSoup(response1.content, "html.parser")

table1 = soup1.find("table")
headers1 = [header.text.strip() for header in table.find_all("th")]
headers1 = headers1[:2]
rows1 = []
state_names1 = []

for row in table1.find_all("tr")[1:]:
    state_name1 = row.find("th").find("a").text.strip()
    state_names1.append(state_name1)
    cells1 = [cell.text.strip() for cell in row.find_all("td")]
    rows1.append(cells1)

df_d = pd.DataFrame(rows1, columns=headers1)
df_d.index = state_names1

In [6]:
df_d = df_d.rename(columns={'Starbucks Stores 2023': 'dunkin_locations'})
df_d = df_d.drop(['State'], axis=1)

In [7]:
df_d['State'] = df_d.index
df_d = df_d.reset_index(drop=True)

In [8]:
df = pd.merge(df_s, df_d, on='State', how='outer')

In [9]:
df_d

dunkin_locations                 State
0             1,414              New York
1             1,068         Massachusetts
2               883               Florida
3               866            New Jersey
4               692              Illinois
5               629          Pennsylvania
6               480           Connecticut
7               303              Maryland
8               265               Georgia
9               245                  Ohio
10              230              Virginia
11              196                 Texas
12              206        North Carolina
13              218         New Hampshire
14              159                 Maine
15              160          Rhode Island
16              134            California
17              123             Tennessee
18              104               Indiana
19              103        South Carolina
20              105              Michigan
21              102               Arizona
22               83             Wisconsin
23               59               Alabama
24               66              Delaware
25               50              Missouri
26               41              Kentucky
27               49               Vermont
28               43              Colorado
29               38                Nevada
30               34             Minnesota
31               34                  Iowa
32               27         West Virginia
33               28                Kansas
34               23              Nebraska
35               18              Oklahoma
36                   District of Columbia
37               19            New Mexico
38               16             Louisiana
39                9              Arkansas
40               12           Mississippi
41                6                Hawaii
42                1                  Utah
43                1               Wyoming
44                0                Alaska
45                0                 Idaho
46                0               Montana
47                0          North Dakota
48                0                Oregon
49                0          South Dakota
50               19            Washington

## Parts 4, 5, 6

In [10]:
response2 = requests.get("https://simple.wikipedia.org/wiki/List_of_U.S._states_by_population")
soup2 = BeautifulSoup(response2.content, "html.parser")
table2 = soup2.find("table", attrs={
    "class": ["wikitable", "sortable"],
    "style": "width:100%; text-align:center;"
})
headers2 = [th.text.strip() for th in table2.find_all("th")]
rows2 = []

for tr in table2.find_all("tr")[1:]:
    cells2 = []
    for td in tr.find_all("td"):
        if td.find("a"):
            cells2.append(td.find("a").text.strip())
        else:
            cells2.append(td.text.strip())
    rows2.append(cells2)

df_table2 = pd.DataFrame(rows2, columns=headers2)

In [11]:
df_combined = pd.merge(df, df_table2, on='State', how='outer')
df_combined = df_combined.loc[:, ['State', 'starbucks_locations', 'dunkin_locations', 'Census population, April 1, 2020[1][2]']]

In [12]:
df_combined = df_combined.rename(columns={'Census population, April 1, 2020[1][2]': 'population'})

In [13]:
df_combined['Dunkin_stock_price'] = 106.48
df_combined['starbucks_stock_price'] = 98.02

In [14]:
regions = {
    'West': ['California', 'Nevada', 'Oregon', 'Washington', 'Hawaii', 'Alaska', 'Arizona', 'Colorado', 'Idaho', 'Montana', 'New Mexico', 'Utah', 'Wyoming'],
    'South': ['Texas', 'Florida', 'Georgia', 'Alabama', 'Delware', 'Maryland', 'North Carolina', 'South Carolina', 'Virginia', 'Kentucky', 'Mississippi', 'Tennessee', 'Arkansas', 'Louisiana', 'Oklahoma'],
    'Northeast': ['New York', 'New Jersey', 'Pennsylvania', 'Connecticut', 'Maine', 'Massachusetts', 'New Hampshire', 'Rhode Isalnd', 'Vermont'],
    'Midwest': ['Illinois,' 'Indiana', 'Michigan', 'Ohio', 'Wisconsin', 'Iowa', 'Kansas', 'Minnesota', 'Missouri', 'Nebraska', 'North Dakota', 'South Dakota']
}

def get_region(state):
    for region, states in regions.items():
        if state in states:
            return region
    return 'Other'

df_combined['region'] = df_combined['State'].apply(get_region)

In [15]:
df_combined

State starbucks_locations  \
0                                       Alabama                  85   
1                                        Alaska                  49   
2                                American Samoa                 NaN   
3                                       Arizona                 548   
4                                      Arkansas                  55   
5                                    California               3,080   
6                                      Colorado                 495   
7                                   Connecticut                 123   
8                      Contiguous United States                 NaN   
9                                      Delaware                  25   
10                         District of Columbia                  91   
11                          Fifty states + D.C.                 NaN   
12                                      Florida                 844   
13                                      Georgia                 326   
14                                         Guam                 NaN   
15                                       Hawaii                  99   
16                                        Idaho                  67   
17                                     Illinois                 677   
18                                      Indiana                 221   
19                                         Iowa                  89   
20                                       Kansas                  94   
21                                     Kentucky                 166   
22                                    Louisiana                  84   
23                                        Maine                  30   
24                                     Maryland                 257   
25                                Massachusetts                 273   
26                                     Michigan                 283   
27                                    Minnesota                 184   
28                                  Mississippi                  32   
29                                     Missouri                 188   
30                                      Montana                  32   
31                                     Nebraska                  58   
32                                       Nevada                 253   
33                                New Hampshire                  29   
34                                   New Jersey                 261   
35                                   New Mexico                  76   
36                                     New York                 692   
37                               North Carolina                 338   
38                                 North Dakota                  13   
39                     Northern Mariana Islands                 NaN   
40                                         Ohio                 491   
41                                     Oklahoma                  79   
42                                       Oregon                 359   
43                                 Pennsylvania                 357   
44                                  Puerto Rico                 NaN   
45                                 Rhode Island                  27   
46                               South Carolina                 131   
47                                 South Dakota                  25   
48                                    Tennessee                 180   
49                                        Texas               1,346   
50                             The fifty states                 NaN   
51  Total U.S. (including D.C. and territories)                 NaN   
52                          U.S. Virgin Islands                 NaN   
53                                         Utah                 101   
54                                      Vermont                   8   
55                                     Virginia                 48

## Part 7 (code below)
- Are some of these chains more prevalent in certain states than others? Possibly despite having less stores overall? Same questions for regions instead of states.

Almost always there are more Dunkin Donuts in the Northeast states, and more Starbucks in the West and South. The Midwest was less consistent. States that stood out were California with 3,080 Starbucks locations and 134 Dunkin Donuts locations, and New York and Massechusetts with 1,414 and 1,068 Dunkins respectively, and 692 and 273 Starbucks respectively.

- Does the distribution of each chain’s stores match population distribution, by both state/region?

Total locations match population distribution, but regionality is important when it comes to distribution for each chain. For example, California has 40 million people and over 3,000 total locations and New York has over 20 millions people and over 2,000 total locations, while Wyoming has 600,000 people and 24 total locations. But even though California has so many people, they only have 134 Dunkin locations.

- Do the financial data match what you’d expect based on the number and locations of the stores? Why or why not?

I chose stock price. This data does not give enough information to make conclusions from number of stores.

## Part 8

In [16]:
def scrape_company_locations(url):
    response = requests.get(url)
    response.raise_for_status()
    soup = BeautifulSoup(response.content, "html.parser")
    table = soup.find("table")
    headers = [header.text.strip() for header in table.find_all("th")]
    headers = headers[:2]
    rows = []
    state_names = []

    for row in table.find_all("tr")[1:]:
        state_name = row.find("th").find("a").text.strip()
        state_names.append(state_name)
        cells = [cell.text.strip() for cell in row.find_all("td")]
        rows.append(cells)

    df = pd.DataFrame(rows, columns=headers)
    df.index = state_names
    df = df.rename(columns={'State': 'locations'})
    df['State'] = df.index
    df = df.reset_index(drop=True)

    return df

In [17]:
scrape_company_locations("https://worldpopulationreview.com/state-rankings/in-n-out-by-state")

locations In-N-Out Locations 2024           State
0        276                     267      California
1         43                      41           Texas
2         35                      34         Arizona
3         23                      22          Nevada
4         13                      12            Utah
5         12                       7        Colorado
6          4                       4          Oregon
7          1                                   Idaho
8          0                                 Alabama
9          0                                  Alaska
10         0                                Arkansas
11         0                             Connecticut
12         0                                Delaware
13         0                                 Florida
14         0                                 Georgia
15         0                                  Hawaii
16         0                                Illinois
17         0                                 Indiana
18         0                                    Iowa
19         0                                  Kansas
20         0                                Kentucky
21         0                               Louisiana
22         0                                   Maine
23         0                                Maryland
24         0                           Massachusetts
25         0                                Michigan
26         0                               Minnesota
27         0                             Mississippi
28         0                                Missouri
29         0                                 Montana
30         0                                Nebraska
31         0                           New Hampshire
32         0                              New Jersey
33         0                              New Mexico
34         0                                New York
35         0                          North Carolina
36         0                            North Dakota
37         0                                    Ohio
38         0                                Oklahoma
39         0                            Pennsylvania
40         0                            Rhode Island
41         0                          South Carolina
42         0                            South Dakota
43         0                               Tennessee
44         0                                 Vermont
45         0                                Virginia
46         0                              Washington
47         0                           West Virginia
48         0                               Wisconsin
49         0                                 Wyoming